Compress PT2E in Executorch

anzr299 opened the pull request:
### Summary
[PLEASE REMOVE] See [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests) for ExecuTorch PR guidelines.

[PLEASE REMOVE] If this PR closes an issue, please add a `Fixes #<issue-id>` line.

[PLEASE REMOVE] If this PR introduces a fix or feature that should be the upcoming release notes, please add a "Release notes: <area>" label. For a list of available release notes labels, check out [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests).

### Test plan
[PLEASE REMOVE] How did you test this PR? Please write down any manual commands you used and note down tests that you have written if applicable.

anzr299 committed:
support nncf compression flow for ov quantizer

anzr299 committed:
Merge branch 'main' into an/openvino/nncf_compress_pt2e

anzr299 committed:
Merge branch 'main' into an/openvino/nncf_compress_pt2e

anzr299 committed:
lint fixes; plus other minor fixes

anzr299 commented on backends/openvino/requirements.txt:
```diff
@@ -1,2 +1,2 @@
 transformers
-git+https://github.com/openvinotoolkit/nncf@3d753ac#egg=nncf
+git+https://github.com/openvinotoolkit/nncf#egg=nncf
```
Yes, that's a good point. I forgot to make it fixed since I assumed it was just develop branch. I will point it to a fixed commit.

anzr299 replied:
I will add the commit after https://github.com/openvinotoolkit/nncf/pull/3782/files is merged

anzr299 commented on backends/openvino/quantization/nncf_compression.py:
```diff
@@ -0,0 +1,79 @@
+def get_calibration_data(
+    module: torch.fx.GraphModule, tokenizer, prompts: str, max_len: int
+):
+    # TODO: change criteria & support batch inputs if necessary
+    pos = torch.tensor(0, dtype=torch.int64)
+    token_list = tokenizer.encode(prompts, bos=True, eos=False)
+
+    with torch.no_grad():
+        while token_list[-1] != tokenizer.eos_id and pos < max_len:
+            logits = module(
+                torch.full((1, 1), token_list[pos]),
+                {"input_pos": torch.tensor((pos,))},
+            )
+            pos += 1
+            if pos >= len(token_list):
+                token_list.append(torch.argmax(logits[:], dim=-1).item())
+    token_list = [
+        (
+            pos,
+            token,
+        )
+        for pos, token in enumerate(token_list)
+    ]
+    return token_list
```
From executorch quantization flow's calibration method: https://github.com/anzr299/executorch/blob/535a19898ee829d939a9d8c762cb89dae9231dd7/extension/llm/export/builder.py#L289

anzr299 replied:
Done

anzr299 commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -23,6 +23,7 @@
 from typing import Callable, List, Optional, Union
 
 import torch
+from executorch.backends.openvino.quantization import apply_nncf_data_aware_compression
```
All other backends directly specify the intiialization in https://github.com/anzr299/executorch/blob/535a19898ee829d939a9d8c762cb89dae9231dd7/examples/models/llama/export_llama_lib.py#L774. 

My intitial approach was to keep the function `apply_nncf_data_aware_compression` in the `export_llama_lib.py` file itself, but mustafa suggested to move it to a seperate file which will sit in our backend directory `executorch/backends/llama` This was as per the suggestions of the executorch team to mustafa. 

If we talk about something similar, where backends like qualcomm have some completely unique implementations for lets say attention, batching and so on. They have created a completely new example and files and stored it all stored in their backend directory as well.

anzr299 replied:
As per our discussion offline, I have made it a lazy import here: https://github.com/anzr299/executorch/blob/d761c3be9c809a78690633393cf0df6e67088cd3/examples/models/llama/export_llama_lib.py#L939

daniil-lyakhov commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -920,9 +936,14 @@ def _to_edge_and_lower_llama_openvino(
-    builder = builder_exported.pt2e_quantize(quantizers).to_edge_transform_and_lower(
-        partitioners
-    )
+    if awq or scale_estimation:
```
I suggest to always use `compress_pt2e`, as we are advertising it as an advanced torch.ao quantization

anzr299 replied:
Done

anzr299 committed:
Merge branch 'main' into an/openvino/nncf_compress_pt2e

anzr299 committed:
fix workflow

anzr299 committed:
fix

anzr299 committed:
fix typehint; access wc algo using absolute

anzr299 commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -119,15 +126,18 @@ def __init__(
+                subset_size=subset_size, **self._wc_config
             )
 
+    def get_wc_algo_configuration(self):
```
This method is used by the adapter in the PR: https://github.com/openvinotoolkit/nncf/pull/3782

anzr299 committed:
make nncf function import conditional

anzr299 commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -124,7 +130,7 @@ def __init__(
             )
             subset_size = 1  # Doesn't really matter in this case since it is data-free. Should just be +ve
-            self._algo = nncf.quantization.algorithms.weight_compression.algorithm.WeightCompression(
+            self._algo = WeightCompression(
```
Done

anzr299 commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -163,6 +169,16 @@ def get_nncf_quantization_setup(
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
+        list[WeightCompressionParameters], Optional[dict[str, WCTensorStatistic]]
```
You're right. Fixed

anzr299 commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -163,6 +169,16 @@ def get_nncf_quantization_setup(
         self._algo._set_backend_entity(model)
         return self._algo.find_quantization_setup(model, nncf_graph)
 
+    def get_nncf_weight_compression_parameters(
+        self,
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
```
This method is used by the openvino adapter

anzr299 commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -241,6 +242,19 @@ def build_args_parser() -> argparse.ArgumentParser:
         help="Path to the adapter_config.json file. Used if the model has trained LoRA adapters. Must provide adapter_checkpoint.",
     )
 
+    parser.add_argument(
+        "--nncf_awq",
+        required=False,
+        action="store_true",
+        help="Whether to use AWQ from NNCF. Applicable only for the Openvino backend.",
+    )
+
+    parser.add_argument(
+        "--nncf_scale_estimation",
+        action="store_true",
+        help="Whether to use Scale Estimation algorithm from NNCF. Applicable only for the Openvino backend",
+    )
+
```
Done

anzr299 commented on backends/openvino/quantization/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .nncf_compression import use_nncf_compression
+
+__all__ = ["use_nncf_compression"]
```
This is to import the function that applies nncf compress pt2e. The name was wrong here. Fixed it

anzr299 committed:
review changes

anzr299 commented on backends/openvino/quantization/nncf_compression.py:
```diff
@@ -0,0 +1,79 @@
+    return token_list
+
+
+def transform_fn(token_pos_map: tuple[int, str]):
```
Done

anzr299 commented on backends/openvino/quantization/nncf_compression.py:
```diff
@@ -0,0 +1,79 @@
+
+def transform_fn(token_pos_map: tuple[int, str]):
+    # tokenized_text = tokenizer.encode(prompts, bos=False, eos=False)
+    inputs = ()
```
Done

daniil-lyakhov commented on backends/openvino/quantizer/llm_compression.py:
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Intel Corporation
```
You sure this llama example specific code is belong to this directory? I'm not sure. How other backends are doing this?

anzr299 replied:
Hm, perhaps I can keep get_calibration_data and move the code from apply_nncf_data_aware_compression and and move the to the example. 

Although, these methods can be used for any LLM which is an object of LLMEdgeManager. This is because get_calibration_data is a common function in executorch and transform_func does not perform any special task

daniil-lyakhov replied:
Let's move this file to the `quantizer/llm_compression.py`

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -119,15 +123,18 @@ def __init__(
                 "wo", ""
             )  # Mode value has to match NNCF CompressWeightsMode
-            weight_compression_configuration = get_weight_compression_configuration(
+            self._wc_config = get_weight_compression_configuration(
```
Let's do `algo.get_config()` and sort outputs to keep only required params there, then we don't have to keep copies in `self._wc_config`

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -119,15 +123,18 @@ def __init__(
             compression_mode = mode.value.replace(
```
Please use dict there

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantization/nncf_compression.py:
```diff
@@ -0,0 +1,87 @@
+
+
+def apply_nncf_data_aware_compression(
+    builder_exported, quantizers, awq: bool, scale_estimation: bool
```
typehints

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -163,6 +170,18 @@ def get_nncf_quantization_setup(
+        self,
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
```
Description please

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -119,15 +123,18 @@ def __init__(
+                subset_size=subset_size, **self._wc_config
             )
 
+    def get_wc_algo_configuration(self):
```
Description

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -119,15 +123,18 @@ def __init__(
             )
 
+    def get_wc_algo_configuration(self):
+        return self._wc_config
```
General comment: try to avoid returning mutable objects as they are, it is always better to return a copy of a dict/list because it saves code from unexpected data modifications

anzr299 replied:
Alright. The config is processed here and only the selected configs which we need are repacked into a dict and returned.

anzr299 committed:
add

anzr299 committed:
add method to obtain the attributes for WC algo which are defined during quantizer intialization;
make the wc config passed to WC algo as a variable in init method instead of class attribute.

anzr299 committed:
use dict to obtain the compression mode

anzr299 committed:
add docstring

anzr299 committed:
review changes

anzr299 committed:
make and move llm_compression file; review changes

anzr299 committed:
revert WC ALgo initialization

daniil-lyakhov commented on backends/openvino/quantizer/llm_compression.py:
```diff
@@ -0,0 +1,103 @@
+    module: torch.fx.GraphModule, tokenizer, prompts: str, max_len: int
+):
+    """
+    This method is used to obtain calibration data from a prompt so that the algorithm
+    is calibrated not only with the dataset but also the inputs which are output-ed by
+    the model.
```
Please specify a model/ models for which this method is applicable

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/llm_compression.py:
```diff
@@ -0,0 +1,103 @@
+
+def transform_fn(token_pos_map: tuple[int, int]):
+    """
+    Transforms and returns input from dataset so that it is acceptable by the model
```
Please specify model/models for which this transform model fn is working

anzr299 replied:
Done

daniil-lyakhov commented on backends/openvino/quantizer/quantizer.py:
```diff
@@ -116,18 +120,35 @@ def __init__(
+    def get_quantizer_specific_internal_algo_config(self) -> Dict[str, Any]:
+        """
+        Returns the config attributes like all_layers, group_size, backup_mode and Quantization mode
+        from the quantizer's internal algorithm. These attributes are defined when initializing the quantizer
```
It is not yet a data class which would specify the interface for compress_pt2e, but let's describe all the params in the output dict. I believe this would help on the code review
```suggestion
        Returns a dictionary with all_layers, group_size, backup_mode and Quantization mode parameters
        used by the compress_pt2e weight compression algorithm. 
        * all_layers - desc...
        * group_size - desc..
        ...
        ```

anzr299 replied:
Done

anzr299 committed:
update name of get compression config method

anzr299 committed:
review changes
